File: Java/src/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
# 
# **********
# Javier Izquierdo Vera 2º B
# Juan Carlos Gallardo Morales 2º B
# **********
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author javieriv
 */
public class Napakalaki {
    // Singleton -> Una instancia únicamente, es lo que devolvemos
    private static final Napakalaki instance = new Napakalaki();

    private int indexCurrentPlayer = -1;
    private ArrayList<Player> players = new ArrayList();
    private Player currentPlayer;
    private Monster currentMonster;
    private CardDealer dealer;
    // Constructor privado para que no se use desde fuera
    private Napakalaki(){
        Player currentPlayer = null;
    	Monster currentMonster = null;
   	CardDealer dealer = CardDealer.getInstance();
    // Constructor privado... singleton
    }
    
    // Devolvemos la única instancia posible
    public static Napakalaki getInstance(){
        return instance;
    }
    
    // Métodos privados
    private void initPlayers(ArrayList<String> names){
        for(int i = 0; i < names.size(); i++)
            players.add(new Player(names.get(i)));
    }
    
    private Player nextPlayer(){
        if (indexCurrentPlayer == -1){
                Random aleatorio = new Random();
                indexCurrentPlayer = aleatorio.nextInt(players.size());
        } else
                indexCurrentPlayer = (indexCurrentPlayer+1)%players.size();
        
        return players.get(indexCurrentPlayer);
    }
    
    private boolean nextTurnAllowed(){
        if (this.currentPlayer == null)
            return true;
        else
            return currentPlayer.validState();
    }
    
    // Métodos públicos
    public CombatResult developCombat(){
        CombatResult cr = currentPlayer.combat(currentMonster);
        dealer.giveMonsterBack(currentMonster);
        
        if (cr == CombatResult.LoseAndConvert){
            CultistPlayer cp = new CultistPlayer(currentPlayer,dealer.nextCultist());
            currentPlayer = (CultistPlayer)cp;
            players.add(indexCurrentPlayer,currentPlayer);
        }
        
        return cr;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        for(int i = 0; i < treasures.size(); i++){
            Treasure treasure = treasures.get(i); // Obtenemos el tesoro
            currentPlayer.discardVisibleTreasure(treasure); // Lo borramos correctamente
            
            dealer.giveTreasureBack(treasure); //Se lo damos al dealer para que lo devuelva al mazo       
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        for(int i = 0; i < treasures.size(); i++){
            Treasure treasure = treasures.get(i); // Obtenemos el tesoro
            currentPlayer.discardHiddenTreasure(treasure); // Lo borramos correctamente
            
            dealer.giveTreasureBack(treasure); //Se lo damos al dealer para que lo devuelva al mazo       
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        for (int i = 0; i < treasures.size(); i++)
            currentPlayer.makeTreasureVisible(treasures.get(i));
    }
    
    public boolean buyLevels(ArrayList<Treasure> visible, ArrayList<Treasure> hidden){
        return currentPlayer.buyLevels(visible, hidden);
    }
    
    public void initGame(ArrayList<String> players){
        this.initPlayers(players);
        dealer = CardDealer.getInstance();
        dealer.initCards();
        this.nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOK = this.nextTurnAllowed();
        boolean dead;
        
        if(stateOK){
            currentMonster = dealer.nextMonster();
            currentPlayer = this.nextPlayer();
            dead = currentPlayer.isDead();
            
            if (dead)
                currentPlayer.initTreasures();

        }
 
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        return (result == CombatResult.WinAndWinGame);
    }
    
    
}

File: Ruby/pkg/Napakalaki-0.0.1/lib/napakalaki.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
# **********
# Javier Izquierdo Vera 2º B
# Juan Carlos Gallardo Morales 2º B
# **********

require 'singleton'
require_relative 'card_dealer'
require_relative 'player'
require_relative 'monster'
require_relative 'combat_result'
require_relative 'CultistPlayer'

module NapakalakiGame

class Napakalaki
  include Singleton
  
  # Get's
  attr_reader:currentPlayer,:currentMonster
  
  def initialize
    @players = Array.new
    @dealer = CardDealer.instance
    @indexCurrentPlayer = -1
    @currentPlayer
    @currentMonster
  end
  # Métodos públicos
  # CombatResult
  def developCombat
    cr = @currentPlayer.combat(@currentMonster)
    @dealer.giveMonsterBack(@currentMonster)
    
    if cr == CombatResult::LOSEANDCONVERT
      @currentPlayer = CultistPlayer.copy(@currentPlayer, @dealer.nextCultist)
      @players[@indexCurrentPlayer] = @currentPlayer
    end
    
    return cr
  end
  
  def discardVisibleTreasures(treasures)
    for t in treasures
        @currentPlayer.discardVisibleTreasure(t)
        @dealer.giveTreasureBack(t)
    end
  end
  
  def discardHiddenTreasures(treasures)
    for t in treasures
        @currentPlayer.discardHiddenTreasure(t)
        @dealer.giveTreasureBack(t)
    end
  end
  
  def makeTreasuresVisible(treasures)
    for trea in treasures
        @currentPlayer.makeTreasureVisible(trea)
    end
  end
  
  #boolean
  def buyLevels(visible, hidden)
     @currentPlayer.buyLevels(visible, hidden)
  end
  
  def initGame(players)
      initPlayers(players)
      @dealer = CardDealer.instance
      @dealer.initCards
      nextTurn
  end
  
  #boolean
  def nextTurnAllowed
    if @currentPlayer == nil
            return true
    else
     @currentPlayer.validState
    end
  end
  
  #boolean
  def endOfGame(result)
    result == CombatResult::WINANDWINGAME
  end
  
  # boolean
  def nextTurn
      stateOK = self.nextTurnAllowed
        
        if stateOK == true
            @currentMonster = @dealer.nextMonster
            @currentPlayer = nextPlayer
            
            dead = @currentPlayer.dead
            
            if dead == true  
              @currentPlayer.initTreasures  
            end
        end
        stateOK
  end
  
  # Métodos privados
  private
  def initPlayers(names)
      for name in names
          @players.push(Player.newPlayer(name))
      end
  end
  
  # Player
  def nextPlayer
      if @indexCurrentPlayer == -1
         @indexCurrentPlayer = rand(@players.size-1)
      else
         @indexCurrentPlayer = (@indexCurrentPlayer + 1) % @players.size
      end
      
      return @players[@indexCurrentPlayer]
  end
  
  
end


end